lib/SILGen/SILGenFunction.cpp

```diff
@@ -380,6 +390,30 @@ SILGenFunction::getOrCreateScope(const ast_scope::ASTScopeImpl *ASTScope,
 
   LLVM_DEBUG( ASTScope->print(llvm::errs(), 0, false, false) );
 
+  // Collect all variable declarations in this scope.
+  struct Consumer : public namelookup::AbstractASTScopeDeclConsumer {
+    const ast_scope::ASTScopeImpl *ASTScope;
+    VarDeclScopeMapTy& VarDeclScopeMap;
+    Consumer(const ast_scope::ASTScopeImpl *ASTScope,
+             VarDeclScopeMapTy &VarDeclScopeMap)
+        : ASTScope(ASTScope), VarDeclScopeMap(VarDeclScopeMap) {}
+
+    bool consume(ArrayRef<ValueDecl *> values,
+                 NullablePtr<DeclContext> baseDC) override {
+      for (auto &value : values)
+        VarDeclScopeMap.insert({value, ASTScope});
```

Comment:
Maybe assert here that we didn't have a duplicate?

Reply:
Done. (And fixed a few missed caching opportunities along the way).